### ISSUE TITLE
Dockstore GH App appears to be inconsistent with what it considers a push

### ISSUE THREAD

**Issue by aofarrel** [dockstore/dockstore#4831]
**Describe the bug**
It appears that the Dockstore GitHub App is not consistent with what it considers a push. This affects people adding new repos to Dockstore when it has access to all repos on a GitHub user. We can extrapolate that this might also affect an app being installed with universal access on a GitHub organization, or maybe even all GitHub app installations in general.

**Case One**
1. (months ago) Install docktore github app on aofarrel so that it has access to all repositories
2. (months ago) Add a few aofarrel repos to Dockstore (covstats, ORFalyzer, etc) via the Dockstore GitHub App
3. (sometime last week) Fork the fragpipe workflows onto aofarrel 
4. (Friday, April 1st) Create a valid .dockstore.yml for fragpipe locally
5. Push that valid .dockstore.yml for fragpipe's four workflows
6. Notice that none of fragpipe's workflows are showing up in Dockstore's unpublished tab
7. Notice that fragpipe **does** show up in the Dockstore GitHub app logs as "success" and "install" at 12:33 PM
8. Be puzzled; decide to go do something else for three hours
9. Shrug, add a space to fragpipe's readme, and push it
10. fragpipe's four repos suddenly appear in Dockstore's unpublished tab
11. Notice that fragpipe now shows up in the Dockstore GitHub app logs with a new entry as "success" and "push" at 15:35 PM

I did some kind of reversing and/or squashing *after* step 11 so the only changes that show up are the addition of the .dockstore.yml in my PR. That is probably why you see more than one "push" event in the github app logs, but that does not explain the three hour time period between the install and the first push wherein none of the fragpipe workflows appeared on Dockstore.

<img width="781" alt="Screen Shot 2022-04-04 at 4 18 49 PM" src="https://user-images.githubusercontent.com/27784612/161649849-50acfa22-ecf5-471a-ba78-6c65685f2f2c.png">

**Consequences of Case One to User**
This contradicts what is currently live on documentation, which indicates that you only need an additional commit after publishing a .dockstore.yml if the GH app was recently installed. It is likely to cause some frustration for users, especially as the logs indicate a successful install during step 7.

**Case Two**
1. [Make a brand new repo of dubious utility](https://github.com/aofarrel/fresh-foo)
2. Push partytime.wdl
3. Push valid .dockstore.yml
4. See the workflow on Dockstore as expected, but notice that the logs flagged the single commit where you pushed a valid .dockstore.yml with three events -- install, push failure, and push success.

It is possible that the push failure is caused by the "malformed" WDL, but if so, why would that be considered another push? It's the same push, it should either be a success or a failure, not two pushes.

**Consequences of Case Two to User**
This one isn't a big deal, since at the end of the day the expected behavior technically did happen. But it is odd, and if a user had an issue with my .dockstore.yml it would have been much more difficult to debug seeing as a valid one can cause a push failure.

<img width="788" alt="Screen Shot 2022-04-04 at 4 17 06 PM" src="https://user-images.githubusercontent.com/27784612/161649865-41dca7c8-2c15-4f4a-97ca-cd3548444691.png">

**Expected behavior**
> 1. Enable GitHub Repo for repo foo
> 2. Add/commit valid .dockstore.yml to repo foo locally on branch goo
> 3. Push branch goo
> 4. The workflow should show up in Dockstore.
(Quoting Charles, but that was my expectation too)

**Screenshots**
See above.

**Additional context**
[Webservice](https://github.com/dockstore/dockstore/releases/tag/1.12.0) - 1.12.0
[UI](https://github.com/dockstore/dockstore-ui2/releases/tag/2.9) - 2.9
[Compose Setup](https://github.com/dockstore/compose_setup/releases/tag/1.12.0-rc.1) - 1.12.0-rc.1
[Deploy](https://github.com/dockstore/dockstore-deploy/releases/tag/1.12.3) - 1.12.3
galaxyParsingLambdaVersion: 0.0.7

**Comment by aofarrel** [comment]
(Will delete this comment once I can edit the above)
I'm being blocked from editing the above, maybe unito-bot is hiccuping, but it should be:

**Describe the bug**
It appears that the Dockstore GitHub App is not consistent with what it considers a push. This affects people adding new repos to Dockstore when it has access to all repos on a GitHub user. We can extrapolate that this might also affect an app being installed with universal access on a GitHub organization, or maybe even all GitHub app installations in general.

**Case One**
1. (months ago) Install docktore github app on aofarrel so that it has access to all repositories
2. (months ago) Add a few aofarrel repos to Dockstore (covstats, ORFalyzer, etc) via the Dockstore GitHub App
3. (sometime last week) Fork the fragpipe workflows onto aofarrel 
4. (Friday, April 1st) Create a valid .dockstore.yml for fragpipe locally
5. Push that valid .dockstore.yml for fragpipe's four workflows
6. Notice that none of fragpipe's workflows are showing up in Dockstore's unpublished tab
7. Notice that fragpipe **does** show up in the Dockstore GitHub app logs as "success" and "install" at 12:33 PM
8. Be puzzled; decide to go do something else for three hours
9. Shrug, add a space to fragpipe's readme, and push it
10. fragpipe's four repos suddenly appear in Dockstore's unpublished tab
11. Notice that fragpipe now shows up in the Dockstore GitHub app logs with a new entry as "success" and "push" at 15:35 PM

I did some kind of reversing and/or squashing *after* step 11 so the only changes that show up are the addition of the .dockstore.yml in my PR. That is probably why you see more than one "push" event in the github app logs, but that does not explain the three hour time period between the install and the first push wherein none of the fragpipe workflows appeared on Dockstore.

<img width="781" alt="Screen Shot 2022-04-04 at 4 18 49 PM" src="https://user-images.githubusercontent.com/27784612/161649849-50acfa22-ecf5-471a-ba78-6c65685f2f2c.png">

**Consequences of Case One to User**
This contradicts what is currently live on documentation, which indicates that you only need an additional commit after publishing a .dockstore.yml if the GH app was recently installed. It is likely to cause some frustration for users, especially as the logs indicate a successful install during step 7.

**Case Two**
To keep things simple this scenario had no squashing, reverting, etc.
1. [Make a brand new repo of dubious utility](https://github.com/aofarrel/fresh-foo)
2. Push partytime.wdl
3. Push valid .dockstore.yml
4. See the workflow on Dockstore as expected, but notice that the logs flagged the single commit where you pushed a valid .dockstore.yml with three events -- install, push failure, and push success.

It is possible that the push failure is caused by the "malformed" WDL...
<img width="857" alt="Screen Shot 2022-04-04 at 4 47 12 PM" src="https://user-images.githubusercontent.com/27784612/161650300-892c4ce6-bf3d-429b-be0f-4b8909ad9578.png">

...but if so, why would that be considered another push? It's the same push, it should either be a success or a failure, not two pushes.

**Consequences of Case Two to User**
This one isn't a big deal, since at the end of the day the expected behavior technically did happen. But it is odd, and if a user had an issue with my .dockstore.yml it would have been much more difficult to debug seeing as a valid one can cause a push failure.

<img width="788" alt="Screen Shot 2022-04-04 at 4 17 06 PM" src="https://user-images.githubusercontent.com/27784612/161649865-41dca7c8-2c15-4f4a-97ca-cd3548444691.png">

**Expected behavior**
> 1. Enable GitHub Repo for repo foo
> 2. Add/commit valid .dockstore.yml to repo foo locally on branch goo
> 3. Push branch goo
> 4. The workflow should show up in Dockstore.
(Quoting Charles, but that was my expectation too)

**Screenshots**
See above.

**Additional context**
[Webservice](https://github.com/dockstore/dockstore/releases/tag/1.12.0) - 1.12.0
[UI](https://github.com/dockstore/dockstore-ui2/releases/tag/2.9) - 2.9
[Compose Setup](https://github.com/dockstore/compose_setup/releases/tag/1.12.0-rc.1) - 1.12.0-rc.1
[Deploy](https://github.com/dockstore/dockstore-deploy/releases/tag/1.12.3) - 1.12.3
galaxyParsingLambdaVersion: 0.0.7

**Comment by coverbeck** [comment]
So, case 2 is not a bug and is behaving as predicted, as far as I can tell.

Case 1 would have to be narrowed down for us to realistically do anything about it.

I stand by my statement that the a second push is not necessary :) . Although I'll qualify it that the first push has to have a valid .dockstore.yml pointing to a valid workflow.

**Comment by aofarrel** [comment]
> > **Case Two**
> > ```
> > 1. [Make a brand new repo of dubious utility](https://github.com/aofarrel/fresh-foo)
> > 
> > 2. Push partytime.wdl
> > 
> > 3. Push valid .dockstore.yml
> > 
> > 4. See the workflow on Dockstore as expected, but notice that the logs flagged the single commit where you pushed a valid .dockstore.yml with three events -- install, push failure, and push success.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It is possible that the push failure is caused by the "malformed" WDL, but if so, why would that be considered another push? It's the same push, it should either be a success or a failure, not two pushes.
> 
> I don't understand this. The number of events makes sense to me, the install event corresponds to when the app is installed. Then in step 2 and 3, there are two pushes. So three events in total which seems ok to me?

Step 2 is a push without a dockstore.yml, so I do not expect it to show up in the GitHub logs. Step 3 is the push where the .dockstore.yml is added. If it is retroactively picking up on the push from step 2, which is not what I am expecting but would be fine, then there's no issue with it case 2 other than it being confusing.

**Comment by aofarrel** [comment]
> So, case 2 is not a bug and is behaving as predicted, as far as I can tell.
> 
> Case 1 would have to be narrowed down for us to realistically do anything about it.
> 
> I stand by my statement that the a second push is not necessary :) . Although I'll qualify it that the first push has to have a valid .dockstore.yml pointing to a valid workflow.

Agreed that case 1 needs to be narrowed down, but the fact of the mater is I had a case where pushing a commit with a valid .dockstore.yml was not enough to get the repo to show up on Dockstore. My concern is that users will run into something similar, so regardless of us being able to narrow it down, I think the docs need to address it (albeit perhaps not as strongly as my PR suggested).

**Comment by denis-yuen** [comment]
> Step 2 is a push without a dockstore.yml, so I do not expect it to show up in the GitHub logs.  

I don't think this is the case. For example, doesn't your screenshot above include a push event without a .dockstore.yml?

**Comment by coverbeck** [comment]
> Agreed that case 1 needs to be narrowed down, but the fact of the matter is I had a case where pushing a commit with a valid .dockstore.yml was not enough to get the repo to show up on Dockstore.

Do you have any details that allows us to look into that? When exactly did that happen and with which repo?

> My concern is that users will run into something similar, so regardless of us being able to narrow it down, I think the docs need to address it (albeit perhaps not as strongly as my PR suggested).

I think it would cause even more confusion and what would it say? If what you say is true, that some pushes don't update, then why would the second push be any more reliable? It would have to cover that as well -- keep pushing until you see your changes?

**Comment by aofarrel** [comment]
> > Step 2 is a push without a dockstore.yml, so I do not expect it to show up in the GitHub logs.
> 
> I don't think this is the case. For example, doesn't your screenshot above include a push event without a .dockstore.yml?

That is in case 1, which occurred on March 31st. I didn't make the repo for case 2 until yesterday (April 4th).

**Comment by aofarrel** [comment]
> > Agreed that case 1 needs to be narrowed down, but the fact of the matter is I had a case where pushing a commit with a valid .dockstore.yml was not enough to get the repo to show up on Dockstore.
> 
> Do you have any details that allows us to look into that? When exactly did that happen and with which repo?

It was March 31st. Can't recall exact times but my recollection of the approximate times align with the logs I screenshot for case 1 -- according to the logs, at 12:33 the install was registered, but it was not until 15:35 after a new push that the apps register a push event. Unfortunately I can't link you to a specific commit as I did some squashing to clean up the PR I was making against the original fragpipe repo, but this is the repo: https://github.com/aofarrel/fragpipe-proteomics-pipeline-tutorial Basically I was making sure my .dockstore.yml was valid before opening a PR for Dave; I did not publicly publish my copies that lead me to seeing this behavior. What I do know is that the .dockstore.yml was valid as soon as I first committed + pushed it -- I did not make any changes to the .dockstore.yml file since noon on March 31st (including any changes that would have gotten squashed).

> 
> > My concern is that users will run into something similar, so regardless of us being able to narrow it down, I think the docs need to address it (albeit perhaps not as strongly as my PR suggested).
> 
> I think it would cause even more confusion and what would it say? If what you say is true, that some pushes don't update, then why would the second push be any more reliable? It would have to cover that as well -- keep pushing until you see your changes?

It does indeed open a can of worms and raise a lot of questions. But I think we can protect users from that discussion by just focusing on what the user experiences: When adding a .dockstore.yml to a repo, sometimes, you need to make an additional commit.

This is technically behavior we have already documented nestled in an FAQ answer (line 75, github-apps-troubleshooting-tips.rst)...

> If access was given to the wrong organization or repository, or this is your first time installing the app, you'll need to push another commit after adding the correct repository to activate the sync to Dockstore.

...all that is changing that we are no longer adding the qualifier "you only have to do this if you just installed the Dockstore GitHub App for the first time." 

How we replace that qualifier will shape what the user expects. In the short term, before investigating what's going on, we can address this in a few ways:
A) Universally tell the user to push an additional commit the first time you add a new .dockstore.yml to a new repo -- this is technically not necessary in all cases, but it will prevent the user from getting tripped up if the mysterious behavior in case 1 happens to them, and it is technically simpler then what we already telling them.
B) Edit the existing FAQ item to remove the qualifier about this only happening for first installs, ie
> Try pushing another commit to activate the sync to Dockstore.
C) A + B

B and C do admittedly raise the questions you are talking about -- the scenarios when another commit are not explained (because we ourselves do not know them). But A avoids that conversation entirely by implying that is always required. That is not the truth, but if it overcomes the bug, is very simple to do, and does not confuse the user... that might be acceptable.

**Comment by coverbeck** [comment]
> It was March 31st. Can't recall exact times but my recollection of the approximate times align with the logs I screenshot for case 1 -- according to the logs, at 12:33 the install was registered, but it was not until 15:35 after a new push that the apps register a push event.

Ah interesting, we were in the middle of a deploy then. The worst case in theory in that case is that it should work, although there might be a delay. Have you seen this any other time?

**Comment by coverbeck** [comment]
I logged a ticket for this, https://ucsc-cgl.atlassian.net/browse/SEAB-4203.

Thanks for helping us find this.

It's still a very specific window where this can occur, maybe 30 minutes total over the last 9 months, so I still don't think this should be documented.

**Comment by aofarrel** [comment]
I usually push several commits at once so I haven't noticed it up until this point.
